python/pyspark/sql/classic/dataframe.py

```diff
@@ -0,0 +1,1974 @@
+                return toJArray(gateway, jclass, parameter)
+            else:
+                return parameter
+
+        jdf = self._jdf.hint(name, self._jseq(parameters, _converter))
+        return DataFrame(jdf, self.sparkSession)
+
+    def count(self) -> int:
+        return int(self._jdf.count())
+
+    def collect(self) -> List[Row]:
+        with SCCallSiteSync(self._sc):
+            sock_info = self._jdf.collectToPython()
+        return list(_load_from_socket(sock_info, BatchedSerializer(CPickleSerializer())))
+
+    def toLocalIterator(self, prefetchPartitions: bool = False) -> Iterator[Row]:
+        with SCCallSiteSync(self._sc):
+            sock_info = self._jdf.toPythonIterator(prefetchPartitions)
+        return _local_iterator_from_socket(sock_info, BatchedSerializer(CPickleSerializer()))
+
+    def limit(self, num: int) -> "ParentDataFrame":
+        jdf = self._jdf.limit(num)
+        return DataFrame(jdf, self.sparkSession)
+
+    def offset(self, num: int) -> "ParentDataFrame":
+        jdf = self._jdf.offset(num)
+        return DataFrame(jdf, self.sparkSession)
+
+    def take(self, num: int) -> List[Row]:
+        return self.limit(num).collect()
+
+    def tail(self, num: int) -> List[Row]:
+        with SCCallSiteSync(self._sc):
+            sock_info = self._jdf.tailToPython(num)
+        return list(_load_from_socket(sock_info, BatchedSerializer(CPickleSerializer())))
+
+    def foreach(self, f: Callable[[Row], None]) -> None:
+        self.rdd.foreach(f)
+
+    def foreachPartition(self, f: Callable[[Iterator[Row]], None]) -> None:
+        self.rdd.foreachPartition(f)  # type: ignore[arg-type]
+
+    def cache(self) -> "ParentDataFrame":
+        self.is_cached = True
+        self._jdf.cache()
+        return self
+
+    def persist(
+        self,
+        storageLevel: StorageLevel = (StorageLevel.MEMORY_AND_DISK_DESER),
+    ) -> "ParentDataFrame":
+        self.is_cached = True
+        javaStorageLevel = self._sc._getJavaStorageLevel(storageLevel)
+        self._jdf.persist(javaStorageLevel)
+        return self
+
+    @property
+    def storageLevel(self) -> StorageLevel:
+        java_storage_level = self._jdf.storageLevel()
+        storage_level = StorageLevel(
+            java_storage_level.useDisk(),
+            java_storage_level.useMemory(),
+            java_storage_level.useOffHeap(),
+            java_storage_level.deserialized(),
+            java_storage_level.replication(),
+        )
+        return storage_level
+
+    def unpersist(self, blocking: bool = False) -> "ParentDataFrame":
+        self.is_cached = False
+        self._jdf.unpersist(blocking)
+        return self
+
+    def coalesce(self, numPartitions: int) -> "ParentDataFrame":
+        return DataFrame(self._jdf.coalesce(numPartitions), self.sparkSession)
+
+    @overload
+    def repartition(self, numPartitions: int, *cols: "ColumnOrName") -> "ParentDataFrame":
+        ...
```

Comment:
I'm wondering if we need `@overload` definitions in the subclasses?

Reply:
I initially added, and remove it back because MyPy complains too much. I will take another look.

Reply:
Seems like by right we should redefine the overloads here (https://github.com/python/mypy/issues/5146, https://github.com/python/mypy/issues/10699). However, we're using `pyspark.sql.DataFrame` type hints even within our codebase .. so I think it's better to don't have them defined here for now.